### PR TITLE
feat!: replace reusable action with composite action

### PR DESCRIPTION
**Pull request by KeisukeShima**

### CODE REVIEW

**Comment by kenji-miyake** [comment]
As this repository provides CI features, I think you should use `feat` for this PR.
Also, since it breaks the existing features, `feat!` is better.

File: generate-docs/mkdocs-requirements.txt [hunk 1]
```diff
@@ -2,7 +2,6 @@ fontawesome_markdown
 markdown
 mdx_truly_sane_lists
 mdx_unimoji
-mike
```
Comment (kenji-miyake) [inline generate-docs/mkdocs-requirements.txt:5]: Why did you remove this?

File: get-docs-version-name/action.yml [hunk 2]
```diff
@@ -0,0 +1,34 @@
+name: "get-docs-version-name"
+description: "Get the version name of the docs"
```
Comment (kenji-miyake) [inline get-docs-version-name/action.yml:2]: Where does this description show up?
IMO, since we won't publish this to GitHub Marketplace, it's better to comment in README.md instead of this description.

File: generate-docs/action.yml [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+      uses: actions/checkout@v2
+      with:
+        # Fetch with depth=0 to push to gh-pages
+        fetch-depth: 0
+
+    - name: Install MkDocs
+      run: |
+        python3 -m pip install --upgrade pip setuptools wheel
+        pip3 install -r ${GITHUB_ACTION_PATH}/mkdocs-requirements.txt
+      shell: bash
```
Comment (kenji-miyake) [inline generate-docs/action.yml:25]: IMO, writing `shell` between `name` and `run` looks clean.

File: colcon-build-and-test/action.yml [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+  rosdistro:
+    description: "ROS distro"
+    required: true
+  build-depends-repos:
+    description: "VCS repositories containing build dependencies"
+    required: true
+  target-packages:
+    description: "Target packages to build and test"
+    required: true
+  github-access-token:
```
Comment (kenji-miyake) [inline colcon-build-and-test/action.yml:14]: [Nits] Why not just `token`? I think it's generally used and enough.

File: colcon-build-and-test/action.yml [hunk 5]
```diff
@@ -0,0 +1,90 @@
+name: "colcon-build-and-test"
```
Comment (kenji-miyake) [inline colcon-build-and-test/action.yml:1]: [Nits] Why did you add double quotes? I'd like to know if there are any specific reasons.

File: colcon-build-and-test/action.yml [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+outputs:
+  compile-commands-hash:
+    description: Calculated hash of compile_commands.json
+    value: ${{ steps.calc-hash.outputs.hash }}
+
+runs:
+  using: "composite"
+  steps:
+    - name: Checkout repository
+      if: ${{ inputs.target-packages != '' }}
```
Comment (kenji-miyake) [inline colcon-build-and-test/action.yml:26]: I think you can remove this check because it can be checked in the caller side.

File: get-modified-source-files/action.yml [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+        # List added/modified/renamed files excluding deleted ones
+        modified_files=$(git --no-pager diff --name-only --diff-filter=d origin/"${{ github.base_ref }}"...HEAD)
+        modified_source_files=$(echo "$modified_files" | egrep "\.(cpp|hpp)$") || true
+        echo ::set-output name=modified-source-files::${modified_source_files}
+      shell: "bash"
+
+    - name: Show result
+      run: |
+        echo "modified source files: ${{ steps.get-modified-source-files.outputs.modified-source-files }}"
+      shell: "bash"
```
Comment (kenji-miyake) [inline get-modified-source-files/action.yml:30]: Let's make it consistent whether we should use `""` or not.
I prefer no quotes.

image

File: generate-docs/action.yml [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+
+    - name: Create comment body
+      if: ${{ github.event.pull_request }}
+      id: create-comment-body
+      run: |
+        mkdocs-yaml=$(find . -type f -path './mkdocs.yaml' -or -path './mkdocs.yml')
+        site_url=$(grep -oP "site_url:\s+\K(.*)" "$mkdocs-yaml")
+        base_url="${site_url}/pr-${{ github.event.pull_request.number }}/"
+        ${GITHUB_ACTION_PATH}/create-doc-comment.sh "${base_url}" "origin/${{ github.event.pull_request.base.ref }}"
```
Comment (kenji-miyake) [inline generate-docs/action.yml:75]: Please change the comment in `create-doc-comment.sh` like this.

`# Create comment body for generate-docs action`

File: get-modified-source-files/action.yml (hunk 7 above)
Comment (kenji-miyake) [reply get-modified-source-files/action.yml:30]: I'd like to introduce `yamllint`.
https://github.com/adrienverge/yamllint
https://yamllint.readthedocs.io/en/stable/rules.html#module-yamllint.rules.quoted_strings

File: clang-tidy/action.yml [hunk 9]
```diff
@@ -0,0 +1,87 @@
+name: "clang-tidy"
```
Comment (kenji-miyake) [inline clang-tidy/action.yml:1]: Please use `.yaml` for YAML extensions.

File: generate-docs/action.yml (hunk 3 above)
Comment (KeisukeShima) [reply generate-docs/action.yml:25]: Thanks for your comment. I prefer to follow the official documentation.
https://docs.github.com/en/actions/creating-actions/creating-a-composite-action

File: colcon-build-and-test/action.yml (hunk 5 above)
Comment (KeisukeShima) [reply colcon-build-and-test/action.yml:1]: I saw this document.
https://docs.github.com/en/actions/creating-actions/creating-a-composite-action

File: colcon-build-and-test/action.yml (hunk 6 above)
Comment (KeisukeShima) [reply colcon-build-and-test/action.yml]: Removed 2479eb6

File: generate-docs/action.yml (hunk 8 above)
Comment (KeisukeShima) [reply generate-docs/action.yml:75]: daf3528

File: generate-docs/action.yml (hunk 3 above)
Comment (kenji-miyake) [reply generate-docs/action.yml:25]: Okay, but please refrain from resolving conversations before the reviewers check it.
It makes it hard to find the comment.

File: clang-tidy/action.yml [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+    required: true
+  github-access-token:
+    description: "If the repository is private, specify a token."
+    required: false
+
+runs:
+  using: "composite"
+  steps:
+    - name: Checkout repository
+      if: ${{ inputs.target-files != '' }}
```
Comment (kenji-miyake) [inline clang-tidy/action.yml:25]: same as https://github.com/autowarefoundation/autoware-github-actions/pull/4#discussion_r763749610

File: colcon-build-and-test/action.yml (hunk 5 above)
Comment (kenji-miyake) [reply colcon-build-and-test/action.yml:1]: I understand, but I believe what is important is consistency, so I'll apply yamllint later.

**Comment by kenji-miyake** [comment]
Seeing some official documents, I feel "no quotes" is the majority.
https://docs.github.com/en/actions/learn-github-actions/understanding-github-actions#create-an-example-workflow
https://docs.github.com/en/actions/creating-actions/creating-a-composite-action#creating-an-action-metadata-file

So I'll add `yamllint` with the settings and make it consistent.

File: clang-tidy/action.yml (hunk 9 above)
Comment (kenji-miyake) [reply clang-tidy/action.yml]: 185a9d97b5ba160c3ae41f4343566af1119dbd6a

File: get-modified-source-files/action.yml (hunk 7 above)
Comment (kenji-miyake) [reply get-modified-source-files/action.yml]: 0aedfe0d88221bea0e9d036b9320a7553e191c3c

File: colcon-build-and-test/action.yml (hunk 5 above)
Comment (kenji-miyake) [reply colcon-build-and-test/action.yml]: 0aedfe0d88221bea0e9d036b9320a7553e191c3c

File: get-docs-version-name/action.yml (hunk 2 above)
Comment (kenji-miyake) [reply get-docs-version-name/action.yml]: I mean, if we write these descriptions, we should make them more clear and easy to understand.
If not, I think it's better to remove these and write good READMEs instead to reduce the maintenance cost.

Here is an example of the good README.
https://github.com/tier4/github-actions/blob/21419e073219963876fab34be75814658c632814/ros-modified-packages/README.md

File: generate-docs/mkdocs-requirements.txt (hunk 1 above)
Comment (KeisukeShima) [reply generate-docs/mkdocs-requirements.txt:5]: Because this action requires the latest mike.

https://github.com/autowarefoundation/autoware-github-actions/pull/4/files#diff-e452dcac3881d99adfa804391ebf04a95794941fa9071c0d0a0c682f4d9f97eeR28-R31

Comment (kenji-miyake) [reply generate-docs/mkdocs-requirements.txt:5]: It's just a temporary workaround.
If there is no problem, please keep this.

File: get-docs-version-name/action.yml (hunk 2 above)
Comment (KeisukeShima) [reply get-docs-version-name/action.yml]: ???
This syntax is required.
https://docs.github.com/en/actions/creating-actions/metadata-syntax-for-github-actions#inputsinput_iddescription

File: generate-docs/action.yml (hunk 3 above)
Comment (KeisukeShima) [reply generate-docs/action.yml]: Yes, basically I don't resolve conversations, but in this case I think it would save you some trouble if I resolved the conversation.

File: get-docs-version-name/action.yml (hunk 2 above)
Comment (kenji-miyake) [reply get-docs-version-name/action.yml]: Okay, I see. I thought it was optional because other workflows are probably so.
Then, we have to fix them so that people can understand.
Also anyway, we need READMEs separately, so could you add READMEs? Empty ones are acceptable for now.

File: clang-tidy/action.yml (hunk 10 above)
Comment (KeisukeShima) [reply clang-tidy/action.yml]: fa09d64

File: get-docs-version-name/action.yml (hunk 2 above)
Comment (kenji-miyake) [reply get-docs-version-name/action.yml]: Or I feel making the description to be the same as the name is acceptable.

File: clang-tidy/action.yml (hunk 10 above)
Comment (kenji-miyake) [reply clang-tidy/action.yml]: Thank you! Then please add the empty check here instead.

https://github.com/autowarefoundation/autoware.universe/pull/104/files#diff-d549e0b4270f5b02caec3d0bf592280fe38c24737b9d34cc9cbee0552ab0773aR20-R26

File: get-docs-version-name/action.yml (hunk 2 above)
Comment (KeisukeShima) [reply get-docs-version-name/action.yml]: > Then, we have to fix them so that people can understand.

I agree with you. The official GitHub documentation is confusing.

> Also anyway, we need READMEs separately, so could you add READMEs? Empty ones are acceptable for now.

It's good to understand for users. I will add README.

File: colcon-build-and-test/action.yml (hunk 4 above)
Comment (KeisukeShima) [reply colcon-build-and-test/action.yml]: OK. Changed 77d3c9b

File: clang-tidy/action.yaml [hunk 11]
```diff
@@ -0,0 +1,78 @@
...
+
+    - name: Show target files
+      run: |
+        echo "target files: ${{ inputs.target-files }}"
+      shell: bash
+
+    - name: Set git config for private repositories
+      run: |
+        git config --local --unset-all http.https://github.com/.extraheader || true
+        git config --global url.https://${{ inputs.token }}@github.com.insteadOf 'https://github.com'
```
Comment (kenji-miyake) [inline clang-tidy/action.yaml:47]: I'm sorry, I forgot to modify this before.
Adding `x-access-token:` is more useful, I remember.

```suggestion
        git config --global url.https://x-access-token:${{ inputs.token }}@github.com.insteadOf 'https://github.com'
```

File: get-docs-version-name/action.yml (hunk 2 above)
Comment (KeisukeShima) [reply get-docs-version-name/action.yml]: Added readme
df9cc4a

File: remove-docs/mkdocs-requirements.txt [hunk 12]
```diff
@@ -0,0 +1,12 @@
+fontawesome_markdown
```
Comment (kenji-miyake) [inline remove-docs/mkdocs-requirements.txt:1]: Same as https://github.com/autowarefoundation/autoware-github-actions/pull/4/files#r763735937

File: .github/workflows/pre-commit.yaml [hunk 13]
```diff
@@ -2,9 +2,31 @@ name: pre-commit
...
+    runs-on: ubuntu-latest
+
+    steps:
+      - name: Checkout repository
+        uses: actions/checkout@v2
+
+      - name: Set pre-commit-config
+        run: |
+          PRE_COMMIT_CONFIG=${{ inputs.pre-commit-config }}
+          echo "PRE_COMMIT_CONFIG=${PRE_COMMIT_CONFIG:=.pre-commit-config.yaml}" >> $GITHUB_ENV
```
Comment (kenji-miyake) [inline .github/workflows/pre-commit.yaml:26]: Why did you add this section? I guess it's not necessary.

Comment (kenji-miyake) [reply .github/workflows/pre-commit.yaml:26]: Ah, it's because we explicitly use `--config`.

Comment (kenji-miyake) [reply .github/workflows/pre-commit.yaml:26]: Mm... should we make pre-commit a composite action? :thinking: 
I'll consider it.

File: clang-tidy/action.yml (hunk 10 above)
Comment (KeisukeShima) [reply clang-tidy/action.yml]: Added https://github.com/autowarefoundation/autoware.universe/pull/104/commits/0759fb96e1752e4c9da52f3de27e63434628fed3

Comment (kenji-miyake) [reply clang-tidy/action.yml]: Does no `!= ''` work correctly? I'm not sure.
Could you refer to the official document (of context variables?) if possible?

Comment (KeisukeShima) [reply clang-tidy/action.yml]: Empty strings will be converted to 0, so I think it will work.
https://docs.github.com/en/actions/learn-github-actions/expressions#operators

**Comment by kenji-miyake** [comment]
Although it's yet WIP, I fixed a bit.
